songbird/src/input/ytdl_src.rs

```diff
@@ -0,0 +1,76 @@
+use super::{
+    child_to_reader,
+    error::{Error, Result},
+    Codec,
+    Container,
+    Input,
+    Metadata,
+};
+use std::process::{Command, Stdio};
+use tracing::debug;
+
+/// Creates a streamed audio source with `youtube-dl` and `ffmpeg`.
+pub async fn ytdl(uri: &str) -> Result<Input> {
+    _ytdl(uri, &[]).await
+}
+
+pub(crate) async fn _ytdl(uri: &str, pre_args: &[&str]) -> Result<Input> {
+    let ytdl_args = [
+        "-f",
+        "webm[abr>0]/bestaudio/best",
+        "-R",
+        "infinite",
+        "--no-playlist",
+        "--ignore-config",
+        uri,
+        "-o",
+        "-",
+    ];
+
+    let ffmpeg_args = [
+        "-f",
+        "s16le",
+        "-ac",
+        "2",
+        "-ar",
+        "48000",
+        "-acodec",
+        "pcm_f32le",
+        "-",
+    ];
+
+    let youtube_dl = Command::new("youtube-dl")
+        .args(&ytdl_args)
+        .stdin(Stdio::null())
+        .stderr(Stdio::null())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    let ffmpeg = Command::new("ffmpeg")
+        .args(pre_args)
+        .arg("-i")
+        .arg("-")
+        .args(&ffmpeg_args)
+        .stdin(youtube_dl.stdout.ok_or(Error::Stdout)?)
+        .stderr(Stdio::null())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    let metadata = Metadata::from_ytdl_uri(uri).await;
```

Comment:
Instead of calling youtube-dl twice, once to pipe to ffmpeg and another to fetch metadata it may be better to use youtube-dl to dump the playlist/url to the stream and metadata at the same time so that only one call is made to the remote site.

Reply:
Is there a natural way of doing this, built into either ffmpeg or ytdl? I understand that ytdl embeds its metadata within the output stream when called normally, but extracting it is another problem:
* Having ytdl tell us what we need requires the "-j" option or similar, preventing any stream output. This led to the current solution.
* ffprobe seems to be required to get in-stream metadata in JSON or other easy-to-parse format from media. This works well with an on-disk input source, but consumes piped output as far as I can tell -- which still needs to go into ffmpeg.

I'd really like it if there's a way to wrangle ffmpeg (or ytdl) to, e.g., put f32 PCM stream output on Stdout and dump (only!) metadata to Stderr, but I'm not nearly familiar enough with them right now to pull that off.

Reply:
You can use `youtube-dl --dump-json` this will give you a dump of all the metadata of the target media. You can then use the `url` field to pass to ffmpeg to play the actual media. You can see how mpv uses this (sort of, it's a lua binding) here https://github.com/mpv-player/mpv/blob/5dcfe32ff21bf560a27db3a75f773f73afdf6dcf/player/lua/ytdl_hook.lua

Reply:
I feel we did this before (i.e., years ago), and ran into all sorts of terrifying DASH issues where ytdl sources would just stop at random points which ffmpeg just gave up on. I'll try `--print-json`, which seems to do what I want.

Reply:
Figured it out now!